refactor!: Move gracefulShutdownTimeout to roleGroup config (#486)
stackabletech/trino-operator: docs/modules/trino/pages/usage-guide/operations/graceful-shutdown.adoc

= Graceful shutdown

You can configure the graceful shutdown as described in xref:concepts:operations/graceful_shutdown.adoc[].

== Coordinators

As a default, coordinators have `15 minutes` to terminate gracefully.

The coordinator process will always run as PID `1` and will get a `SIGTERM` once Kubernetes wants to terminate the Pod.
It will initiate a graceful shutdown.

> After the graceful shutdown timeout is passed and the process still didn't exit, Kubernetes will issue an `SIGKILL` to force-kill the process.

Reviewer: ```suggestion
After the graceful shutdown timeout runs out, and the process still didn't exit, Kubernetes will issue a `SIGKILL` signal to force-kill the process.
```

> When a coordinator gets restarted all running queries will fail and will not be recovered.

Reviewer: ```suggestion
When a coordinator gets restarted, all currently running queries will fail and cannot be recovered after the restart process is finished.
```

> As of Trino version `429` this can not be prevented (e.g. by using multiple coordinators).

Reviewer: ```suggestion
Starting with Trino version `429` this behavior can not be prevented (e.g. by using multiple coordinators).
```

Author: It actually was never supported. We just put in the version 428/429 here to be still correct, even when e.g. version 450 adds support for that.

== Workers

> As a default, Coordinators have `60 minutes` to terminate gracefully.

Reviewer: Add a newline.

> Trino supports https://trino.io/docs/current/admin/graceful-shutdown.html[graceful shutdown] of the workers.

Reviewer: ```suggestion
Trino supports https://trino.io/docs/current/admin/graceful-shutdown.html[gracefully shutting down] workers.
```

No additional configuration is needed, this guide is intended for users that need to tweak this mechanism.

The default graceful shutdown period is `1` hour, but it can be configured as follows:

[source,yaml]
----
apiVersion: trino.stackable.tech/v1alpha1
kind: TrinoCluster
metadata:
  name: trino
spec:
  # ...
  workers:
    config:
      gracefulShutdownTimeout: 1h
    roleGroups:
      default:
        replicas: 1
----

== Implementation
Once a worker Pod is asked to terminate, the `PreStop` hook is executed and the following timeline occurs:

1. The worker goes into `SHUTTING_DOWN` state.
2. The worker sleeps for `30` seconds to ensure that the coordinator has noticed the shutdown and stops scheduling new tasks on the worker.
3. The worker now waits till all tasks running on it complete. This will take as long as the longest running query takes.
4. The worker sleeps for `30` seconds to ensure that the coordinator has
noticed that all tasks are complete
5. The `PreStop` hook will never return, but the JVM will be shut down by the graceful shutdown mechanism.

> 6. When the graceful shutdown is not quick enough (e.g. a query runs longer than the graceful shutdown period), after `<graceful shutdown period> + 30s of step 2 + 30s of step 4 + 10s safety overhead` the Pod gets killed, regardless if it has shut down gracefully or not. This is achieved by setting `terminationGracePeriodSeconds` on the worker Pods. In this case queries running on the worker will fail.

Reviewer: ```suggestion
6. If the graceful shutdown doesn't complete quick enough (e.g. a query runs longer than the graceful shutdown period), after `<graceful shutdown period> + 30s of step 2 + 30s of step 4 + 10s safety overhead` the Pod gets killed, regardless if it has shut down gracefully or not. This is achieved by setting `terminationGracePeriodSeconds` on the worker Pods. Currently running queries on the worker will fail and cannot be recovered.
```

> WARNING: As of version 23.7, the secret-operator issues TLS certificates with a lifetime of 24h. It also adds an annotation to the Pod, so that it is restarted 30 minutes before the certificate expires (23.5h hours in this case). Currently, this results in all Pod using https (both coordinator and workers in a typical setup) restarting every 23.5 hours.

Reviewer: ```suggestion
WARNING: Starting with SDP version `23.7`, the secret-operator issues TLS certificates with a lifetime of 24h. It also adds an annotation to the Pod, indicating it requires a restart 30 minutes before the certificate expires (23.5h hours in this case). Currently, this results in all Pod using HTTPS (both coordinator and workers in a typical setup) to be restarted every 23.5 hours.
```

Author: Please have a look at a217d95547c756f427fc3b6919f1ba95b0d201d0

The TLS certificate lifetime can be configured once https://github.com/stackabletech/secret-operator/pull/306 is merged.

== Implications

> All queries that take less than the minimal graceful shutdown period of all roleGroups (`1` hour as a default) are guaranteed to not be disturbed by regular termination of Pods.

Reviewer: Add a newline.

> They can obviously still fail when e.g. a Kubernetes node dies completely or the Pod does not get the time it takes to properly gracefully shut down.

Reviewer: ```suggestion
They can obviously still fail when, for example, a Kubernetes node dies or the graceful shutdown timeout configured on the Pod is not long enough.
```

Author: Changed to `They can obviously still fail when, for example, a Kubernetes node dies completely or the Pod does not get the time it takes to properly gracefully shut down - e.g. the Kubernetes node gets rebooted before it is fully drained.`

Reviewer: This version is slightly more compact: `They can obviously still fail when, for example, a Kubernetes node dies or gets rebooted before it is fully drained.`

Author: Changed in 41596a4

> Because of this the operator automatically restricts the execution time of queries to the minimal graceful shutdown period of all roleGroups using the Trino configuration `query.max-execution-time=3600s`.

Reviewer: ```suggestion
Because of this, the operator automatically restricts the execution time of queries to the minimal graceful shutdown period of all roleGroups using the Trino configuration `query.max-execution-time=3600s`.
```

== Implementation

Once a worker Pod is asked to terminate, the `PreStop` hook is executed and the following timeline occurs:

1. The worker goes into `SHUTTING_DOWN` state.
2. The worker sleeps for `30` seconds to ensure that the coordinator has noticed the shutdown and stops scheduling new tasks on the worker.
3. The worker now waits till all tasks running on it complete. This will take as long as the longest running query takes.
4. The worker sleeps for `30` seconds to ensure that the coordinator has
noticed that all tasks are complete
5. The `PreStop` hook will never return, but the JVM will be shut down by the graceful shutdown mechanism.
6. If the graceful shutdown doesn't complete quick enough (e.g. a query runs longer than the graceful shutdown period), after `<graceful shutdown period> + 30s of step 2 + 30s of step 4 + 10s safety overhead` the Pod gets killed, regardless if it has shut down gracefully or not. This is achieved by setting `terminationGracePeriodSeconds` on the worker Pods. Currently running queries on the worker will fail and cannot be recovered.

[WARNING]
====
As of SDP version `23.7`, the secret-operator issues TLS certificates with a lifetime of 24h.

> It also adds an annotation to the Pod, indicting it requires a restart 30 minutes before the certificate expires (23.5h hours in this case).

Reviewer: ```suggestion
It also adds an annotation to the Pod, indicating it requires a restart 30 minutes before the certificate expires (23.5h hours in this case).
```

Typo.

        query.max-execution-time: 24h
----

Please keep in mind, that queries taking longer than the graceful shutdown period are now subject to failure when a Trino worker gets shut down.

> This can be circumvented by using https://trino.io/docs/current/admin/fault-tolerant-execution.html[Fault-tolerant execution], which is not supported natively yet.

Reviewer: ```suggestion
Running into this issue can be circumvented by using https://trino.io/docs/current/admin/fault-tolerant-execution.html[Fault-tolerant execution], which is not supported natively yet.
```

> Until properly supported, you have to use configOverrides to enable it.

Reviewer: ```suggestion
Until native support is added, you will have to use `configOverrides` to enable it.
```
